Generate Feature-Blame-Report

sisteu56 opened the pull request.

Commit:
implement FeatureBlameReportExperiment

Commit:
add test_repo to test projects

Commit:
some nmae/type fixes

Commit:
correct binary?

Commit:
added print feature

Commit:
change config

Commit:
create bc with clang

Commit:
print out FBR

Commit:
remove unnecessary code

Commit:
minor cleanup

Commit:
added new file

Commit:
fixes for pre-commit hook

boehmseb commented on varats/varats/projects/cpp_projects/commit_feature_interactions_test_repo.py:
```diff
@@ -0,0 +1,53 @@
+    SOURCE = [
+        PaperConfigSpecificGit(
+            project_name="CFI",
+            remote="https://github.com/sisteu56/cfi_test_repo.git",
```
We should integrate this into our test repo infrastructure.
@sisteu56 we can discuss this in our next meeting.

boehmseb commented on varats/varats/projects/cpp_projects/commit_feature_interactions_test_repo.py:
```diff
@@ -0,0 +1,53 @@
+    test_files = ["main.cpp", "helper.cpp"]
+
+    @property
+    def binaries(self) -> tp.List[ProjectBinaryWrapper]:
```
You should override the function `binaries_for_revision` instead of this one.
This function has the advantage that it can declare binaries per revision.

PS: could you tell me where you have copied this from? We should fix that location so that it isn't used as a reference in the future.

sisteu56 replied:
I copied it from here https://vara.readthedocs.io/en/vara-dev/vara-ts-api/projects.html#projects

boehmseb commented on varats/varats/data/reports/feature_blame_report.py:
```diff
@@ -0,0 +1,271 @@
+    def __init__(self, feature: str, commit_hash: str, repo: str) -> None:
+        self.__feature = feature
+        self.__commit_hash = commit_hash
+        self.__repo = repo
```
Use `CommitRepoPair` from `git_util.py` instead.

boehmseb commented on varats/varats/data/reports/feature_blame_report.py:
```diff
@@ -0,0 +1,271 @@
+    ) -> 'FeatureCommitRegion':
+        """Creates a 'FeatureCommitRegion' from the corresponding yaml document
+        section."""
+        feature = raw_region_data["feature"]
```
On the C++ side you have a vector of features and store it in the yaml with the key `features`. Here you only expect one feature with the key `feature` instead.

sisteu56 replied:
That's because I changed the FeatureCommitRegion class on the C++ side last week, but forgot to push it

boehmseb commented on varats/varats/data/reports/feature_blame_report.py:
```diff
@@ -0,0 +1,271 @@
+        print("    -FEATURE COMMIT REGION INSTRUCTION")
+        print("      -INSTRUCTION: " + self.__instruction)
+        print("      -LOCATION: " + self.__location)
+        for FeatureCommitRegion in self.__feature_commit_regions:
```
You cannot name a variable the same as an existing class. Use `feature_commit_region` instead.
Note that the naming conventions for variables in Python (`snake_case`) are different from the ones in VaRA/LLVM (`PascalCase`).

boehmseb commented on varats/varats/data/reports/feature_blame_report.py:
```diff
@@ -0,0 +1,271 @@
+        """'FeatureBlameResultFunctionEntry' prints itself."""
+        print("  -FEATURE BLAME RESULT FUNCTION ENTRY")
+        print("    -FUNCTION: " + self.demangled_name)
+        for FeatureCommitRegionInstruction in self.__feature_commit_region_insts:
```
Same as above

boehmseb commented on varats/varats/data/reports/feature_blame_report.py:
```diff
@@ -0,0 +1,271 @@
+    def print(self) -> None:
+        """'FeatureBlameReport' prints itself."""
+        print("FEATURE BLAME REPORT")
+        for FeatureBlameResultFunctionEntry in self.__function_entries.values():
```
Same as above.

boehmseb commented on varats/varats/data/reports/feature_blame_report.py:
```diff
@@ -0,0 +1,271 @@
+        """
+        return self.__function_entries[mangled_function_name]
+
+        # def get_feature_locations_dict(self) -> tp.Dict[str, tp.Set[str]]:
```
Always comment out the whole function to avoid type errors. This includes the docstring since this is also a statement in python.

Commit:
review changes

Commit:
fixes by pre-commit

Commit:
Merge branch 'vara-dev' of github.com:se-sic/VaRA-Tool-Suite into f-GenerateFeatureBlameReport

Commit:
adapt code to work after pull from vara-dev

Commit:
adapted python fbr to new yaml fbr

Commit:
separated feature blame report
Structural- and DataflowFeatureBlameReport can be generated separatedly now

boehmseb commented on varats/varats/experiments/vara/feature_blame_experiment.py:
```diff
@@ -0,0 +1,73 @@
+"""Implements the base feature blame experiment, making it easier to create
```
This is all very similar to the module `blame_experiment.py`.
I suggest merging these two and renaming the module to something like `vara_experiments.py`.
@vulder what's your take on this?

sisteu56 replied:
Should I rename generate_basic_feature_blame_experiment_actions and generate_basic_blame_experiment_actions into one function generate_basic_vara_experiment_actions or something like this?

Commit:
Merge branch 'vara-dev' of github.com:se-sic/VaRA-Tool-Suite into f-GenerateFeatureBlameReport

Commit:
added detection of feature model

Commit:
refactoring

Commit:
Merge branch 'vara-dev' of github.com:se-sic/VaRA-Tool-Suite into f-GenerateFeatureBlameReport

Commit:
changes for checks

Commit:
Merge branch 'vara-dev' of github.com:se-sic/VaRA-Tool-Suite into f-GenerateFeatureBlameReport

Commit:
added missing FeatureModelProvider to dataflow report

Commit:
adapt test_bb_config for merge of experiments
perviously merged blame_experiment and feature_blame_experiment into one file vara_experiments

boehmseb commented on varats/varats/experiments/vara/vara_experiments.py:
```diff
@@ -51,6 +52,37 @@ def setup_basic_blame_experiment(
+        experiment.get_handle(), project, report_type
+    )
+
+    # This c-flag is provided by VaRA and it suggests to use the git-blame
+    # annotation.
```
```suggestion
    # These flags are provided by VaRA and suggest to use git-blame
    # and feature annotations.
```

boehmseb commented on varats/varats/experiments/vara/vara_experiments.py:
```diff
@@ -71,3 +103,25 @@ def generate_basic_blame_experiment_actions(
     )
+
+
+def generate_basic_feature_blame_experiment_actions(
```
That's the same as `generate_basic_blame_experiment_actions`, right?
In that case, you should reuse the existing function.

boehmseb commented on varats/varats/data/reports/feature_blame_report.py:
```diff
@@ -0,0 +1,190 @@
+        """commit of this cfi."""
+        return self.__commit
+
+    def is_terminator(self) -> bool:
+        br_regex = re.compile(r'(br( i1 | label ))|(switch i\d{1,} )')
+        return br_regex.search(self.__feature) is not None
```
What do you need this for? I don't think this is a reliable method to identify terminator instructions.

sisteu56 replied:
It's not needed, it was just left from copying functions from the FeatureAnalysisReport. 
I will remove it

boehmseb commented on varats/varats/data/reports/feature_blame_report.py:
```diff
@@ -0,0 +1,190 @@
+        """commit of this cfi."""
+        return self.__commits
+
+    def is_terminator(self) -> bool:
+        br_regex = re.compile(r'(br( i1 | label ))|(switch i\d{1,} )')
+        return br_regex.search(self.__feature) is not None
```
Same as above.

boehmseb reviewed (changes requested):
Almost there. A couple of things to clean up left.

Commit:
changes to review

Commit:
remove unused import

Commit:
adapt to changes in structural cfi collection
some refactoring

Commit:
fix mistake in return type
